storage/BloomFilterIndexSubBlock.cpp

```diff
@@ -0,0 +1,271 @@
+  DCHECK(is_initialized_);
+  if (!is_consistent_) {
+    return false;
+  }
+  bool didAllSucceed = true;
+  for (const tuple_id &tuple : tuples) {
+    didAllSucceed = didAllSucceed && addEntry(tuple);
+  }
+  return didAllSucceed;
+}
+
+void BloomFilterIndexSubBlock::removeEntry(const tuple_id tuple) {
+  is_consistent_ = false;
+}
+
+void BloomFilterIndexSubBlock::bulkRemoveEntries(const TupleIdSequence &tuples) {
+  is_consistent_ = false;
+}
+
+predicate_cost_t BloomFilterIndexSubBlock::estimatePredicateEvaluationCost(
+    const ComparisonPredicate &predicate) const {
+  DCHECK(is_initialized_);
+  BloomFilterSelectivity selectivity = getSelectivityForPredicate(predicate);
+  // Note: A Bloomfilter index is only useful when it gives a zero selectivity
+  //       in which case a block can be skipped entirely.
+  if (selectivity == BloomFilterSelectivity::kSelectivityNone) {
+    // Return minimum cost so that this block can be skipped altogether.
+    return predicate_cost::kConstantTime;
+  } else {
+    // Return maximum cost because this is as worse as doing a full scan of the block.
+    return predicate_cost::kInfinite;
+  }
+}
+
+TupleIdSequence* BloomFilterIndexSubBlock::getMatchesForPredicate(
+    const ComparisonPredicate &predicate,
+    const TupleIdSequence *filter) const {
+  DCHECK(is_initialized_);
+  if (filter != nullptr) {
+    LOG(FATAL) << "BloomFilterIndex does not support filter evaluation with predicate.";
+  }
+  BloomFilterSelectivity selectivity = getSelectivityForPredicate(predicate);
+  if (selectivity == BloomFilterSelectivity::kSelectivityNone) {
+    // Best-case scenario, when a BloomFilter index allows to skip a block entirely.
+    // A new tuple ID sequence is initialized to false for all values.
+    return new TupleIdSequence(tuple_store_.numTuples());
+  } else if (selectivity == BloomFilterSelectivity::kSelectivityAll) {
+    TupleIdSequence* tuple_sequence = new TupleIdSequence(tuple_store_.numTuples());
+    // Set all existing tuples to true, selected.
+    if (tuple_store_.isPacked()) {
+      tuple_sequence->setRange(0, tuple_store_.numTuples(), true);
+    } else {
+      for (tuple_id tid = 0; tid <= tuple_store_.getMaxTupleID(); ++tid) {
+        if (tuple_store_.hasTupleWithID(tid)) {
+          tuple_sequence->set(tid, true);
+        }
+      }
+    }
+    return tuple_sequence;
+  } else {
+    LOG(FATAL) << "BloomFilterIndex should never have been called.";
+    return nullptr;
+  }
+}
+
+BloomFilterIndexSubBlock::BloomFilterSelectivity
+    BloomFilterIndexSubBlock::getSelectivityForPredicate(const ComparisonPredicate &predicate) const {
+  DCHECK(is_initialized_);
+  if (!is_consistent_) {
+    return BloomFilterSelectivity::kSelectivityUnknown;
+  }
+  // Bloom filter index can be currently only used to evaluate
+  // equality matching comparison predicates with static values.
+  if (predicate.getComparison().getComparisonID() == ComparisonID::kEqual
```

Comment:
No, currently bloom filter only support static values on the right. Thanks, have updated the comment to indicate that.